### DISCUSSION TITLE
ELI5: tweepy.errors.Unauthorized: 401 Unauthorized 32 - Could not authenticate you

### DISCUSSION THREAD

**Discussion by reagle** [tweepy/tweepy#2120] [Questions] [answered]
I have a simple app that scrapes and posts to Twitter (code below), perhaps 3 or 4 times a week. It was suspended a few days ago -- though I'm not sure why. I know there's different versions of the API, and different authentication mechanisms, and all kind of Musk changes, but is it possible to have a simple app using tweepy still do these things for free?

https://github.com/reagle/thunderdell/blob/main/utils/web.py#L132

https://github.com/reagle/thunderdell/blob/main/formats/scrape/twitter.py

**Comment by Harmon758** [comment, marked as answer]
Create Tweets as with the former, yes.
Retrieve Tweets as with the latter, no.
See https://twittercommunity.com/t/announcing-new-access-tiers-for-the-twitter-api/188728 and https://developer.twitter.com/en/docs/twitter-api/getting-started/about-twitter-api.

**Reply by reagle** [reply]
Is there a recent tutorial for how to do the former then? That is, which API, which OATH mechnanism, etc. I had working code but it was suspended and following existing tutorials isn't working.

**Reply by Harmon758** [reply]
You need to use Twitter API v2 with Free access, so you can use the [POST /2/tweets endpoint](https://developer.twitter.com/en/docs/twitter-api/tweets/manage-tweets/api-reference/post-tweets) with [`Client.create_tweet`](https://tweepy.readthedocs.io/en/v4.14.0/client.html#tweepy.Client.create_tweet).

If by "OATH mechnanism", you mean OAuth / authentication method, you need to authenticate as the user to Tweet as, so you can use OAuth 1.0a User Context or OAuth 2.0 Authorization Code Flow with PKCE (User Context).

There's [examples of both authentication and Tweet creation in the documentation](https://tweepy.readthedocs.io/en/v4.14.0/examples.html).

**Reply by reagle** [reply]
Thank you. But when I do the following I get:  "401 Unauthorized".

```
import tweepy

bearer_token = TW_BEARER_TOKEN

consumer_key = TW_API_KEY
consumer_secret = TW_API_SECRET

access_token = TW_CLIENT_ID
access_token_secret = TW_CLIENT_SECRET

client = tweepy.Client(
    consumer_key=consumer_key,
    consumer_secret=consumer_secret,
    access_token=access_token,
    access_token_secret=access_token_secret,
)

response = client.create_tweet(
    text="This Tweet was Tweeted using Tweepy and Twitter API v2!"
)
print(f"https://twitter.com/user/status/{response.data['id']}")
```

I can do other things, such as see mentions and get recent tweets with this authentication.

**Reply by Harmon758** [reply]
You likely haven't given your app write permissions.
https://developer.twitter.com/en/docs/apps/app-permissions

Otherwise, you'll need to provide the full traceback.

**Reply by reagle** [reply]
Thank you for your patience! The environment for the app is set to "Development" and it is set to read and write.

<img width="1125" alt="SCR-20230602-jped" src="https://github.com/tweepy/tweepy/assets/596814/b18e95ef-340b-4d88-99cb-55f751fc5c95">

**Reply by Harmon758** [reply]
Did you regenerate your credentials after enabling write app permissions, and are you using those regenerated credentials?

**Reply by reagle** [reply]
I believe so, I double-checked when setting it up, but mapping the tokens' names from what the twitter-dev documentation uses and what this documentation uses can be confusing (e.g., client secret is access token secret). Yesterday, I was able to do some other things with the API, so I figured those were correct. I'll have to step away for a bit and try again from the start.

**Reply by Harmon758** [reply]
The client secret is not the access token secret, and the client ID is not the access token.
You're using the wrong credentials.

https://tweepy.readthedocs.io/en/v4.14.0/authentication.html
https://developer.twitter.com/en/docs/twitter-api/getting-started/getting-access-to-the-twitter-api
https://developer.twitter.com/en/docs/authentication/oauth-1-0a
https://developer.twitter.com/en/docs/authentication/oauth-1-0a/api-key-and-secret

Those other methods are probably working because they're using endpoints that don't require you to authenticate as a user. You likely aren't even using OAuth 1.0a with them.
